File: skills/agentcore/scripts/lib/fetcher.py
"""
HTTP fetcher with TTL-based on-disk cache.

Responsible for all network I/O in the skill CLI:
    fetch_raw_url  Fetches raw content (used for llms.txt, README, boto3 index, FAQ pages)
    fetch_doc_page Fetches HTML doc pages, converts to Markdown, caches with TTL

Cache behavior:
    - On-disk files under tempfile.gettempdir()/agentcore-skill-cache (survives between CLI
      runs, since each list/search/fetch invocation is a separate short-lived process)
    - TTL configurable via AGENTCORE_CACHE_TTL_MINUTES (default: 60)
    - Expired entries re-fetched transparently on next access

Standard library only — no third-party dependencies.
"""

import hashlib
import html as html_module
import json
import os
import re
import tempfile
import time
import urllib.request

DEFAULT_TTL_MS = 60 * 60 * 1000
CACHE_DIR = os.path.join(tempfile.gettempdir(), "agentcore-skill-cache")


def _get_cache_ttl_ms():
    env_ttl = os.environ.get("AGENTCORE_CACHE_TTL_MINUTES")
    if env_ttl:
        try:
            return int(env_ttl) * 60 * 1000
        except ValueError:
            pass
    return DEFAULT_TTL_MS


def _now_ms():
    return int(time.time() * 1000)


def _cache_path(url):
    digest = hashlib.sha1(url.encode("utf-8")).hexdigest()
    return os.path.join(CACHE_DIR, digest + ".json")


def _read_cache(url):
    try:
        with open(_cache_path(url), "r", encoding="utf-8") as fh:
            entry = json.load(fh)
        if _now_ms() - entry["timestamp"] < _get_cache_ttl_ms():
            return entry["content"]
    except (OSError, ValueError, KeyError):
        pass
    return None


def _write_cache(url, content):
    try:
        os.makedirs(CACHE_DIR, exist_ok=True)
        with open(_cache_path(url), "w", encoding="utf-8") as fh:
            json.dump({"content": content, "timestamp": _now_ms()}, fh)
    except OSError:
        # caching is best-effort
        pass


def clear_cache():
    import shutil

    try:
        shutil.rmtree(CACHE_DIR, ignore_errors=True)
    except OSError:
        pass


def fetch_raw_url(url):
    """Fetch raw content from a URL. Follows redirects, times out at 15s."""
    req = urllib.request.Request(url, headers={"User-Agent": "AgentCore-Skill/1.0"})
    with urllib.request.urlopen(req, timeout=15) as res:
        charset = res.headers.get_content_charset() or "utf-8"
        return res.read().decode(charset, errors="replace")


def html_to_markdown(html):
    """Convert raw HTML to readable Markdown.

    Extracts the main content area and strips navigation/scripts.
    """
    content = html

    main_match = re.search(
        r'<div id="main-col-body"[^>]*>([\s\S]*?)</div>\s*<div', content
    )
    if main_match:
        content = main_match.group(1)
    else:
        article_match = re.search(r"<main[^>]*>([\s\S]*?)</main>", content)
        if article_match:
            content = article_match.group(1)

    content = re.sub(r"<script[\s\S]*?</script>", "", content, flags=re.IGNORECASE)
    content = re.sub(r"<style[\s\S]*?</style>", "", content, flags=re.IGNORECASE)
    content = re.sub(r"<h1[^>]*>([\s\S]*?)</h1>", r"\n# \1\n", content, flags=re.IGNORECASE)
    content = re.sub(r"<h2[^>]*>([\s\S]*?)</h2>", r"\n## \1\n", content, flags=re.IGNORECASE)
    content = re.sub(r"<h3[^>]*>([\s\S]*?)</h3>", r"\n### \1\n", content, flags=re.IGNORECASE)
    content = re.sub(r"<h4[^>]*>([\s\S]*?)</h4>", r"\n#### \1\n", content, flags=re.IGNORECASE)
    content = re.sub(
        r"<pre[^>]*><code[^>]*>([\s\S]*?)</code></pre>",
        r"\n```\n\1\n```\n",
        content,
        flags=re.IGNORECASE,
    )
    content = re.sub(r"<pre[^>]*>([\s\S]*?)</pre>", r"\n```\n\1\n```\n", content, flags=re.IGNORECASE)
    content = re.sub(r"<code[^>]*>([\s\S]*?)</code>", r"`\1`", content, flags=re.IGNORECASE)
    content = re.sub(r"<strong[^>]*>([\s\S]*?)</strong>", r"**\1**", content, flags=re.IGNORECASE)
    content = re.sub(r"<b[^>]*>([\s\S]*?)</b>", r"**\1**", content, flags=re.IGNORECASE)
    content = re.sub(r"<em[^>]*>([\s\S]*?)</em>", r"*\1*", content, flags=re.IGNORECASE)
    content = re.sub(r"<li[^>]*>([\s\S]*?)</li>", r"- \1\n", content, flags=re.IGNORECASE)
    content = re.sub(
        r'<a[^>]*href="([^"]*)"[^>]*>([\s\S]*?)</a>', r"[\2](\1)", content, flags=re.IGNORECASE
    )
    content = re.sub(r"<p[^>]*>([\s\S]*?)</p>", r"\n\1\n", content, flags=re.IGNORECASE)
    content = re.sub(r"<br\s*/?>", "\n", content, flags=re.IGNORECASE)
    content = re.sub(r"<[^>]+>", "", content)
    content = html_module.unescape(content)
    content = re.sub(r"\n{3,}", "\n\n", content)
    content = content.strip()

    return content


def fetch_doc_page(url):
    """Fetch a documentation page, convert to Markdown, and cache.

    Returns cached content if within TTL, otherwise re-fetches.
    """
    cached = _read_cache(url)
    if cached is not None:
        return cached

    html = fetch_raw_url(url)
    markdown = html_to_markdown(html)
    _write_cache(url, markdown)
    return markdown

File: src/fetcher.ts
/**
 * HTTP fetcher with TTL-based in-memory cache.
 *
 * Responsible for all network I/O in the MCP server:
 *   - fetchRawUrl: Fetches raw content (used for llms.txt, README, boto3 index, FAQ pages)
 *   - fetchDocPage: Fetches HTML doc pages, converts to Markdown, caches with TTL
 *
 * Cache behavior:
 *   - In-memory Map<url, {content, timestamp}>
 *   - TTL configurable via AGENTCORE_CACHE_TTL_MINUTES (default: 60)
 *   - Expired entries re-fetched transparently on next access
 *   - Cache lost on process restart (intentional — guarantees freshness)
 */

import https from "node:https";
import http from "node:http";

interface CacheEntry {
  content: string;
  timestamp: number;
}

const cache = new Map<string, CacheEntry>();
const DEFAULT_TTL_MS = 60 * 60 * 1000;

function getCacheTTL(): number {
  const envTtl = process.env.AGENTCORE_CACHE_TTL_MINUTES;
  if (envTtl) {
    return parseInt(envTtl, 10) * 60 * 1000;
  }
  return DEFAULT_TTL_MS;
}

export function clearCache(): void {
  cache.clear();
}

/**
 * Fetch raw content from a URL. Follows redirects, times out at 15s.
 */
export function fetchRawUrl(url: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const client = url.startsWith("https") ? https : http;
    const req = client.get(url, { headers: { "User-Agent": "AgentCore-Assistant/3.0" } }, (res) => {
      if (res.statusCode && res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        fetchRawUrl(res.headers.location).then(resolve).catch(reject);
        return;
      }
      if (res.statusCode && res.statusCode >= 400) {
        reject(new Error(`HTTP ${res.statusCode} for ${url}`));
        return;
      }
      const chunks: Buffer[] = [];
      res.on("data", (chunk) => chunks.push(chunk));
      res.on("end", () => resolve(Buffer.concat(chunks).toString("utf-8")));
      res.on("error", reject);
    });
    req.on("error", reject);
    req.setTimeout(15000, () => {
      req.destroy();
      reject(new Error(`Timeout fetching ${url}`));
    });
  });
}

/**
 * Convert raw HTML to readable Markdown.
 * Extracts the main content area and strips navigation/scripts.
 */
function htmlToMarkdown(html: string): string {
  let content = html;

  const mainMatch = content.match(/<div id="main-col-body"[^>]*>([\s\S]*?)<\/div>\s*<div/);
  if (mainMatch) {
    content = mainMatch[1];
  } else {
    const articleMatch = content.match(/<main[^>]*>([\s\S]*?)<\/main>/);
    if (articleMatch) content = articleMatch[1];
  }

  content = content.replace(/<script[\s\S]*?<\/script>/gi, "");
  content = content.replace(/<style[\s\S]*?<\/style>/gi, "");
  content = content.replace(/<h1[^>]*>([\s\S]*?)<\/h1>/gi, "\n# $1\n");
  content = content.replace(/<h2[^>]*>([\s\S]*?)<\/h2>/gi, "\n## $1\n");
  content = content.replace(/<h3[^>]*>([\s\S]*?)<\/h3>/gi, "\n### $1\n");
  content = content.replace(/<h4[^>]*>([\s\S]*?)<\/h4>/gi, "\n#### $1\n");
  content = content.replace(/<pre[^>]*><code[^>]*>([\s\S]*?)<\/code><\/pre>/gi, "\n```\n$1\n```\n");
  content = content.replace(/<pre[^>]*>([\s\S]*?)<\/pre>/gi, "\n```\n$1\n```\n");
  content = content.replace(/<code[^>]*>([\s\S]*?)<\/code>/gi, "`$1`");
  content = content.replace(/<strong[^>]*>([\s\S]*?)<\/strong>/gi, "**$1**");
  content = content.replace(/<b[^>]*>([\s\S]*?)<\/b>/gi, "**$1**");
  content = content.replace(/<em[^>]*>([\s\S]*?)<\/em>/gi, "*$1*");
  content = content.replace(/<li[^>]*>([\s\S]*?)<\/li>/gi, "- $1\n");
  content = content.replace(/<a[^>]*href="([^"]*)"[^>]*>([\s\S]*?)<\/a>/gi, "[$2]($1)");
  content = content.replace(/<p[^>]*>([\s\S]*?)<\/p>/gi, "\n$1\n");
  content = content.replace(/<br\s*\/?>/gi, "\n");
  content = content.replace(/<[^>]+>/g, "");
  content = content.replace(/&lt;/g, "<");
  content = content.replace(/&gt;/g, ">");
  content = content.replace(/&amp;/g, "&");
  content = content.replace(/&quot;/g, '"');
  content = content.replace(/&#39;/g, "'");
  content = content.replace(/&nbsp;/g, " ");
  content = content.replace(/\n{3,}/g, "\n\n");
  content = content.trim();

  return content;
}

/**
 * Fetch a documentation page, convert to Markdown, and cache.
 * Returns cached content if within TTL, otherwise re-fetches.
 */
export async function fetchDocPage(url: string): Promise<string> {
  const ttl = getCacheTTL();

  const cached = cache.get(url);
  if (cached && (Date.now() - cached.timestamp) < ttl) {
    return cached.content;
  }

  const html = await fetchRawUrl(url);
  const markdown = htmlToMarkdown(html);

  cache.set(url, { content: markdown, timestamp: Date.now() });

  return markdown;
}
